libs/wingc/src/type_check.rs

```diff
@@ -12,7 +13,70 @@ use type_env::TypeEnv;
 
 use self::type_env::StatementIdx;
 
+pub struct UnsafeRef<T>(*const T);
+impl<T> Clone for UnsafeRef<T> {
+	fn clone(&self) -> Self {
+		Self(self.0)
+	}
+}
+
+impl<T> Copy for UnsafeRef<T> {}
+
+impl<T> std::ops::Deref for UnsafeRef<T> {
+	type Target = T;
+	fn deref(&self) -> &Self::Target {
+		unsafe { &*self.0 }
+	}
+}
+
+impl<T> std::ops::DerefMut for UnsafeRef<T> {
+	fn deref_mut(&mut self) -> &mut Self::Target {
+		unsafe { &mut *(self.0 as *mut T) }
+	}
+}
+
+impl<T> Display for UnsafeRef<T>
+where
+	T: Display,
+{
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		let t: &T = self;
+		write!(f, "{}", t)
+	}
+}
+
+pub type TypeRef = UnsafeRef<Type>;
+
 #[derive(Debug)]
+pub enum IdentKind {
```

Comment:
Wouldn't this technically be `ReferenceKind`?

Reply:
It's a kind of [Identifier](https://en.wikipedia.org/wiki/Identifier_(computer_languages)) or symbol or (named) reference. They are all kind of interchangeable. We can change the name, I'm aiming for something that'll capture the idea of "something with a name". Currently, btw, we have 3 such things: namesapces, custom types, variables/fields.

Reply:
It seemed to mostly match up with the Reference concept we already have in the compiler which is why I had the suggestion. Another option would be `SymbolKind` which I think aligns with your framing as well.

Reply:
I'm more for symbol or identifier because I feel they are more explicit about what this is. For example a namespace's name doesn't feel so much like a reference. 
[Looking at how golang calls this](https://github.com/golang/example/tree/master/gotypes#objects) they call it an "Object" and specifically note:
"_This is different from the everyday meaning of object in object-oriented programming._" I think they were aiming at the subject->object terms used in natural language grammar. The name is the subject and the **object** is the thing. But that felt way too confusing to me.

WDYT about also changing `TypeEnv` / _type_env.rs_ accordingly: `SymbolEnv` / _symbol_env.rs_?

Reply:
I am very much in favor of SymbolEnv